### ISSUE TITLE
docutils<0.18 pin makes rstfmt incompatible with most distros

### ISSUE THREAD

**Issue by jayvdb** [dzhu/rstfmt#25]
https://github.com/dzhu/rstfmt/commit/6025bd2d7d59cbe79b0cf2ccd990c603ded50584 added this pin, and when I try with docutils 0.19 I encounter the following which is probably the problem

```py
[   77s] + python3.8 -m rstfmt --check README.rst sample.rst
[   79s] Traceback (most recent call last):
[   79s]   File "/usr/lib64/python3.8/runpy.py", line 194, in _run_module_as_main
[   79s]     return _run_code(code, main_globals, None,
[   79s]   File "/usr/lib64/python3.8/runpy.py", line 87, in _run_code
[   79s]     exec(code, run_globals)
[   79s]   File "/home/abuild/rpmbuild/BUILD/rstfmt-0.0.11/rstfmt/__main__.py", line 104, in <module>
[   79s]     main()
[   79s]   File "/home/abuild/rpmbuild/BUILD/rstfmt-0.0.11/rstfmt/__main__.py", line 95, in main
[   79s]     do_file(args, path, misformatted)
[   79s]   File "/home/abuild/rpmbuild/BUILD/rstfmt-0.0.11/rstfmt/__main__.py", line 44, in do_file
[   79s]     output = rstfmt.format_node(args.width, doc)
[   79s]   File "/home/abuild/rpmbuild/BUILD/rstfmt-0.0.11/rstfmt/rstfmt.py", line 734, in format_node
[   79s]     ret = "\n".join(fmt(node, FormatContext(0, width, "", "", [], 0)))
[   79s]   File "/home/abuild/rpmbuild/BUILD/rstfmt-0.0.11/rstfmt/rstfmt.py", line 481, in document
[   79s]     yield from chain_intersperse("", fmt_children(node, ctx))
[   79s]   File "/home/abuild/rpmbuild/BUILD/rstfmt-0.0.11/rstfmt/rstfmt.py", line 75, in chain_intersperse
[   79s]     yield from x
[   79s]   File "/home/abuild/rpmbuild/BUILD/rstfmt-0.0.11/rstfmt/rstfmt.py", line 477, in section
[   79s]     yield from chain_intersperse("", fmt_children(node, ctx.in_section()))
[   79s]   File "/home/abuild/rpmbuild/BUILD/rstfmt-0.0.11/rstfmt/rstfmt.py", line 75, in chain_intersperse
[   79s]     yield from x
[   79s]   File "/home/abuild/rpmbuild/BUILD/rstfmt-0.0.11/rstfmt/rstfmt.py", line 437, in paragraph
[   79s]     yield from wrap_text(ctx.width, chain(fmt_children(node, ctx)))
[   79s]   File "/home/abuild/rpmbuild/BUILD/rstfmt-0.0.11/rstfmt/rstfmt.py", line 166, in wrap_text
[   79s]     items = list(items)
[   79s]   File "/home/abuild/rpmbuild/BUILD/rstfmt-0.0.11/rstfmt/rstfmt.py", line 588, in Text
[   79s]     yield node.rawsource.replace(r"\ ", "")
[   79s] AttributeError: 'Text' object has no attribute 'rawsource'
```

Most recent distros are using 0.19 ; c.f. https://repology.org/project/python:docutils/versions

**Comment by jayvdb** [comment]
The following sort of fixes it
```diff
diff --git a/rstfmt/rstfmt.py b/rstfmt/rstfmt.py
index 62bc1dc..0ead7f6 100755
--- a/rstfmt/rstfmt.py
+++ b/rstfmt/rstfmt.py
@@ -585,7 +585,10 @@ class Formatters:
             # this text from adjacent inline markup, but in that case it will be replaced by the
             # wrapping algorithm. Other backslashes may be unnecessary (e.g., "a\` b" or "a\b"), but
             # finding all of those is future work.
-            yield node.rawsource.replace(r"\ ", "")
+            if hasattr(node, "rawsource"):
+                yield node.rawsource.replace(r"\ ", "")
+            else:
+                yield node.astext().replace(r"\ ", "")
         else:
             yield node.astext()
 
```

but then the tests fail with
```python
> make test
PYTHONPATH=. python3 -m rstfmt --check README.rst sample.rst
PYTHONPATH=. python3 -m rstfmt --test README.rst sample.rst
black --check .
All done! ✨ 🍰 ✨
8 files would be left unchanged.
find tests -name '*.rst' -print0 | PYTHONPATH=. xargs -0 python3 -m rstfmt --test -v
============================================================ tests/admonitions.rst
- document {}
    - directive {'directive': <rstfmt.rst_extras.rstfmt_Admonition object at 0x7fb4cd9adba0>}
    - directive {'directive': <rstfmt.rst_extras.rstfmt_Attention object at 0x7fb4cd9ac730>}
    - directive {'directive': <rstfmt.rst_extras.rstfmt_Caution object at 0x7fb4cd9adc90>}
    - directive {'directive': <rstfmt.rst_extras.rstfmt_Danger object at 0x7fb4cd9ac8e0>}
    - directive {'directive': <rstfmt.rst_extras.rstfmt_Error object at 0x7fb4cd9ac880>}
    - directive {'directive': <rstfmt.rst_extras.rstfmt_Hint object at 0x7fb4cd9aceb0>}
    - directive {'directive': <rstfmt.rst_extras.rstfmt_Important object at 0x7fb4cd9ad9c0>}
    - directive {'directive': <rstfmt.rst_extras.rstfmt_Note object at 0x7fb4cd9acfa0>}
    - directive {'directive': <rstfmt.rst_extras.rstfmt_Tip object at 0x7fb4cd9acdc0>}
    - directive {'directive': <rstfmt.rst_extras.rstfmt_Warning object at 0x7fb4cd9ac550>}
============================================================ tests/escaped-text.rst
- document {}
    - paragraph {}
        - Text '- a'
    - paragraph {}
        - Text '#. a'
    - paragraph {}
        - Text 'a\\b'
    - paragraph {}
        - Text 'a'
    - paragraph {}
        - Text 'ab'
    - paragraph {}
        - Text 'a`b'
    - paragraph {}
        - Text 'a `b'
    - paragraph {}
        - Text 'a'
        - literal {}
            - Text 'b'
    - paragraph {}
        - Text 'ab'
    - paragraph {}
        - Text 'a` b'
    - paragraph {}
        - Text 'a` b'
    - paragraph {}
        - Text 'a ` b'
    - paragraph {}
        - Text 'a b`'
:2: (WARNING/2) Bullet list ends without a blank line; unexpected unindent.
Traceback (most recent call last):
  File "/usr/lib64/python3.10/runpy.py", line 196, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/usr/lib64/python3.10/runpy.py", line 86, in _run_code
    exec(code, run_globals)
  File "/home/jayvdb/python/rstfmt/rstfmt/__main__.py", line 104, in <module>
    main()
  File "/home/jayvdb/python/rstfmt/rstfmt/__main__.py", line 95, in main
    do_file(args, path, misformatted)
  File "/home/jayvdb/python/rstfmt/rstfmt/__main__.py", line 39, in do_file
    debug.run_test(doc)
  File "/home/jayvdb/python/rstfmt/rstfmt/debug.py", line 84, in run_test
    doc2 = rstfmt.parse_string(output)
  File "/home/jayvdb/python/rstfmt/rstfmt/rstfmt.py", line 753, in parse_string
    parser.parse(s, doc)
  File "/usr/lib/python3.10/site-packages/docutils/parsers/rst/__init__.py", line 185, in parse
    self.statemachine.run(inputlines, document, inliner=self.inliner)
  File "/usr/lib/python3.10/site-packages/docutils/parsers/rst/states.py", line 169, in run
    results = StateMachineWS.run(self, input_lines, input_offset,
  File "/usr/lib/python3.10/site-packages/docutils/statemachine.py", line 233, in run
    context, next_state, result = self.check_line(
  File "/usr/lib/python3.10/site-packages/docutils/statemachine.py", line 445, in check_line
    return method(match, context, next_state)
  File "/usr/lib/python3.10/site-packages/docutils/parsers/rst/states.py", line 1283, in bullet
    self.parent += self.unindent_warning('Bullet list')
  File "/usr/lib/python3.10/site-packages/docutils/parsers/rst/states.py", line 432, in unindent_warning
    return self.reporter.warning('%s ends without a blank line; '
  File "/usr/lib/python3.10/site-packages/docutils/utils/__init__.py", line 223, in warning
    return self.system_message(self.WARNING_LEVEL, *args, **kwargs)
  File "/home/jayvdb/python/rstfmt/rstfmt/rstfmt.py", line 274, in system_message
    msg = super().system_message(level, message, *children, **kwargs)
  File "/usr/lib/python3.10/site-packages/docutils/utils/__init__.py", line 196, in system_message
    raise SystemMessage(msg, level)
docutils.utils.SystemMessage: :2: (WARNING/2) Bullet list ends without a blank line; unexpected unindent.
make: *** [Makefile:17: test] Error 123
```

**Comment by jayvdb** [comment]
Other than that test failure, it seems to be working OK.  See reformatting of three *.rst in https://github.com/Nuitka/Nuitka/pull/2023

**Comment by jayvdb** [comment]
fwiw, I've packaged this for openSUSE https://build.opensuse.org/package/show/home:jayvdb:py-new/python-rstfmt

**Comment by dzhu** [comment]
This should be fixed in the new 0.0.13 release. Thanks for investigating! Your patch helped as a starting point, though the [final fix](https://github.com/dzhu/rstfmt/commit/0255422f0952ea084bf4b22a4c0964bfccbb0954) ended up being a little different.
